### PR TITLE
feat: (AR-1553 and AR-1649) Splitting search query for federated search + refactor of use cases

### PR DESCRIPTION
**Pull request by dmkasperski**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

This is logic needed in the android reload repo for the federated search also during implementing this in the first place I run into few logical problem so I resolved them and I did planned refactor/improvements of the code 

https://github.com/wireapp/wire-android-reloaded/pull/671

### Solutions

- Create a RegEx to understand when the user is inputting a federated search (`something@something.something`)
- In the `SearchUsersUseCase`, check if the input matches the RegEx and apply federated search

#### Test Coverage

- [X] I have added automated test to this contribution

----
#### PR Post Merge Checklist for internal contributors

 - [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
https://wearezeta.atlassian.net/browse/AR-1553
https://wearezeta.atlassian.net/browse/AR-1649
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/id/QualifiedId.kt [hunk 1]
```diff
@@ -18,6 +18,7 @@ data class QualifiedID(
 }
 
 const val VALUE_DOMAIN_SEPARATOR = "@"
+val FEDERATION_REGEX = Regex("[^@.]+@[^@.]+\\.[^@]+")
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/id/QualifiedId.kt:21]: Are we 100% sure all domains have a dot (`.`) on them?

Also, can we instead use raw strings? So it's easier to copy and paste the pure RegEx without needing to escape them. For example with the `\` inside of it.

```kotlin
val FEDERATION_REGEX = """[^@.]+@[^@.]+\.[^@]+""".toRegex()
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/publicuser/SearchUsersUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        }, {
+            Result.Success(it)
+        })
+    }
+
+
+}
+
+
+
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/publicuser/SearchUsersUseCase.kt:59]: nit: extra lines

```suggestion

}

```

**Review by vitorhugods — COMMENTED** [review]
Nice!

Just tiny suggestions/nit-picks. And I just would like to check that we are sure that Wire domains always contain dots.


Also, in the future, please clean-up the PR description, and checking the boxes, so it gets easier and more straight-forward for reviewers.

**Review by yamilmedina — COMMENTED** [review]
Apart from the comments already added, nice to see the federated search in place during the demo you presented. Nice work @dmkasperski  👏 🪄 

My only comments before merging would be, I think there was a @Deprecated annotation we discussed including, and I didn't see it on the "older" use cases

**Review by yamilmedina — APPROVED** [review]

**Pull request merged**